README.md
# responsive_design

Create responsive applications using different widgets that adapt to different screen sizes.

## Getting Started
[![pub package](https://pub.dev/static/img/pub-dev-logo-2x.png?hash=umitaheu8hl7gd3mineshk2koqfngugi)](https://pub.dev/packages/responsive_design)

Add dependency on your pubspec.yaml file
```yaml
dependencies:
  responsive_design:
```

## **Important!**
Since version `0.1.0` the `ResponsiveSettings` class is no longer used to manage change points.
Instead use `ScreenChangePoints`, check example for more details.

### **Add your own settings**
Manage global the responsive screen change points with `ResponsiveSettings` class.
```dart
void main() {
  // Global Settings
  ResponsiveSettings.instance.setCustomSettings(
    screenChangePoints: const ScreenChangePoints(
      desktopChangePoint: 1150.0,
      tabletChangePoint: 625.0,
      watchChangePoint: 250.0,
    ),
  );
  runApp(const MyApp());
}
```
### **Use the responsive widget**
The responsive widget allows you to have different layouts for each device.
```dart
const ResponsiveWidget(  
  // If don't set the global values will be applied, if do not exist global values,
  // the default values will be applied.
  screenChangePoints: ScreenChangePoints(),
  desktop: Scaffold(
    body: Center(
      child: Text(
        'Desktop screen',
        style: TextStyle(
          color: Colors.black,
          fontSize: 25.0,
        ),
      ),
    ),
  ),
  phone: Scaffold(
    body: Center(
      child: Text(
        'Phone screen',
        style: TextStyle(
          color: Colors.amber,
          fontSize: 25.0,
        ),
      ),
    ),
  ),
  tablet: Scaffold(
    body: Center(
      child: Text(
        'Tablet screen',
        style: TextStyle(
          color: Colors.greenAccent,
          fontSize: 25.0,
        ),
      ),
    ),
  ),
  watch: Scaffold(
    body: Center(
      child: Text(
        'Watch screen',
        style: TextStyle(
          color: Colors.blueAccent,
          fontSize: 25.0,
        ),
      ),
    ),
  ),
);
```
![Responsive Widget](example/assets/images/responsive-widget.gif)

### **Use only responsive app bar**
The responsive app bar allows to adapt the top tool bar to all screens.
```dart
const Scaffold(
  appBar: ResponsiveAppBar(
    // If don't set the global values will be applied, if do not exist global values,
    // the default values will be applied.
    screenChangePoints: ScreenChangePoints(),
    title: Text('Responsive app bar'),
    actions: [
      AppBarAction(
        Center(child: Text('About')),
      ),
      AppBarAction(
        SizedBox(width: 16.0),
        showInAllScreens: true,
      ),
      AppBarAction(
        Icon(Icons.brightness_auto_outlined),
        showInAllScreens: true,
      ),
      AppBarAction(
        SizedBox(width: 16.0),
        showInAllScreens: true,
      ),
    ],
  ),
);
```
![Responsive Widget](example/assets/images/responsive-app-bar.gif)

CHANGELOG.md
## 0.2.5
* Add more widgets to `ResponsiveScaffold`.

## 0.2.4
* Change `sidebar` from `Sidebar` to `Widget` in `ResponsiveScaffold`.

## 0.2.3
* Change `items` from `SidebarItem` to `Widget`.

## 0.2.2
* Add `onDrawerChanged` to `ResponsiveScaffold`.

## 0.2.1
* Add `color`, `selectedColor` and `selected`property to `SidebarItem`.

## 0.2.0
* Fix `elevation` in `ResponsiveAppBar`.
* Add `Sidebar` widget.
* Add `ResponsiveScaffold` widget.

## 0.1.5
* Fix `AppBarAction` space in `ResponsiveAppBar`

## 0.1.4
* Fix center `AppBarAction` on tablet and watch screens.

## 0.1.3
* Export `ResponsiveUtil` class.

## 0.1.2
* Change `changePoints` to `screenChangePoints` for more readable in `ResponsiveSettings` class.

## 0.1.1
* Manage the maximun size of the screen in `ResponsiveSettings`.

## 0.1.0
* Add `ScreenChangePoints` class.
* Now you can settings screen change points globally.

## 0.0.23
* Fix `AppBarAction`

## 0.0.22
* Add the center bool value to `AppBarAction`
* Add space between actions in `ResponsiveAppBar`

## 0.0.21
* Remove `ResponsiveAppBar` padding when use `maxWidth` property.

## 0.0.20
* Add `ResponsiveAppBar` class.
* Add responsive app bar example.

## 0.0.19
* Fix example.

## 0.0.18
* This versión only contains a widget that manage the 4 screen sizes (desktop, phone, tablet, watch).
* Many changes make, please check the example to update to this version.

## 0.0.17
* Add auto implement leading in Responsive App Bar

## 0.0.16
* Remove static unused variables in responsive

## 0.0.15
* Use responsive to show in three screens

## 0.0.14
* Responsive App Bar elevation bug fixed.

## 0.0.13
* Now you can set the gradient background to scaffold

## 0.0.12
* Now you can set the max width of content in responsive scaffold

## 0.0.10
* Now you can set the max width of content in responsive app bar

## 0.0.9
* showTitleInMediumScreen and showTitleInSmallScreen is added on ResponsiveAppBar

## 0.0.8
* Add background color to ResponsiveScaffold

## 0.0.7
* Now you can set the height of ResponsiveAppBar

## 0.0.6
* Now the body of the Scaffold can be extended behind the App Bar

## 0.0.5
* Update example
* Create an Auto Drawer with app bar actions

## 0.0.4
* Update example

## 0.0.3
* Update documentation

## 0.0.2
* No major changes.

## 0.0.1
* Initial release

example/lib/main.dart
import 'package:example/src/sidebar_example.dart';
import 'package:flutter/material.dart';
import 'package:responsive_design/responsive_design.dart';

/// Main function.
void main() {
  // Global Settings
  ResponsiveSettings.instance.setCustomSettings(
    screenChangePoints: const ScreenChangePoints(
      desktopChangePoint: 1150.0,
      tabletChangePoint: 625.0,
      watchChangePoint: 250.0,
    ),
  );
  runApp(const MyApp());
}

/// Root of the application.
class MyApp extends StatelessWidget {
  /// Define a [MyApp] widget.
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Responsive Design Example',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
        appBarTheme: const AppBarTheme(
          backgroundColor: Colors.blueGrey,
        ),
      ),
      routes: {
        '/': (_) => const HomePage(),
        '/responsive-app-bar': (_) => const ResponsiveAppBarExample(),
        '/responsive-widget': (_) => const ResponsiveWidgetExample(),
        '/sidebar-widget': (_) => const SidebarExample(),
      },
      initialRoute: '/',
    );
  }
}

/// Home page of the app.
class HomePage extends StatelessWidget {
  /// Define a [HomePage] widget.
  const HomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: ListView(
          children: [
            ListTile(
              onTap: () => Navigator.of(context).pushNamed(
                '/responsive-app-bar',
              ),
              title: const Text('Responsive App Bar'),
            ),
            ListTile(
              onTap: () => Navigator.of(context).pushNamed(
                '/responsive-widget',
              ),
              title: const Text('Responsive Widget'),
            ),
            ListTile(
              onTap: () => Navigator.of(context).pushNamed(
                '/sidebar-widget',
              ),
              title: const Text('Sidebar Widget'),
            ),
          ],
        ),
      ),
    );
  }
}

/// Example of a responsive app bar.
class ResponsiveAppBarExample extends StatelessWidget {
  /// Define a [ResponsiveAppBarExample] widget.
  const ResponsiveAppBarExample({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const Scaffold(
      appBar: ResponsiveAppBar(
        screenChangePoints: ScreenChangePoints(),
        title: Text('Responsive app bar'),
        actions: [
          AppBarAction(
            Center(child: Text('About')),
          ),
          AppBarAction(
            SizedBox(width: 16.0),
            showInAllScreens: true,
          ),
          AppBarAction(
            Icon(Icons.brightness_auto_outlined),
            showInAllScreens: true,
          ),
          AppBarAction(
            SizedBox(width: 16.0),
            showInAllScreens: true,
          ),
        ],
      ),
    );
  }
}

/// A responsive widget with multiple design for each screen.
class ResponsiveWidgetExample extends StatelessWidget {
  /// Define a [ResponsiveWidgetExample] widget.
  const ResponsiveWidgetExample({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const ResponsiveWidget(
      desktop: Scaffold(
        body: Center(
          child: Text(
            'Desktop screen',
            style: TextStyle(
              color: Colors.black,
              fontSize: 25.0,
            ),
          ),
        ),
      ),
      phone: Scaffold(
        body: Center(
          child: Text(
            'Phone screen',
            style: TextStyle(
              color: Colors.amber,
              fontSize: 25.0,
            ),
          ),
        ),
      ),
      tablet: Scaffold(
        body: Center(
          child: Text(
            'Tablet screen',
            style: TextStyle(
              color: Colors.greenAccent,
              fontSize: 25.0,
            ),
          ),
        ),
      ),
      watch: Scaffold(
        body: Center(
          child: Text(
            'Watch screen',
            style: TextStyle(
              color: Colors.blueAccent,
              fontSize: 25.0,
            ),
          ),
        ),
      ),
    );
  }
}
